sinabs/backend/dynapcnn/sinabs_edges_handler.py

```diff
@@ -0,0 +1,472 @@
+        Will be updated in-place.
+    nodes_io_shapes (dict): Map from node ID to dict containing node's in- and output shapes
+    """
+    # Make sure both source (neuron layer) and target (weight layer) have been previously processed
+    try:
+        source_layer_idx = node_2_layer_map[edge[0]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[0])
+    try:
+        destination_layer_idx = node_2_layer_map[edge[1]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[1])
+
+    # Add new destination
+    output_shape = nodes_io_shapes[edge[0]]["output"]
+    layer_info = dynapcnn_layer_info[source_layer_idx]
+    layer_info["destinations"].append(
+        {
+            "pooling_ids": [],
+            "pooling_modules": [],
+            "destination_layer": destination_layer_idx,
+            "output_shape": output_shape,
+        }
+    )
+
+    # Add output shape of this layer to input shapes of destination
+    dynapcnn_layer_info[destination_layer_idx]["inferred_input_shapes"].add(
+        output_shape
+    )
+
+
+def set_pooling_layer_destination(
+    dynapcnn_layer_info: Dict[int, Dict],
+    edge: Edge,
+    node_2_layer_map: Dict[int, int],
+    nodes_io_shapes: Dict[int, Dict[str, Tuple[Size, Size]]],
+) -> None:
+    """Set destination layer with pooling.
+
+    Parameters
+    ----------
+    dynapcnn_layer_info: Dict with one entry for each future dynapcnn layer.
+        key is unique dynapcnn layer ID, value is dict with nodes of the layer
+        Will be updated in-place.
+    edge: Tuple of 2 integers, indicating edge between two nodes in graph.
+        Edge source has to be within an existing entry of `dynapcnn_layer_info`.
+    node_2_layer_map (dict): Maps each node ID to the ID of the layer it is assigned to.
+        Will be updated in-place.
+    nodes_io_shapes (dict): Map from node ID to dict containing node's in- and output shapes
+    """
+    # Make sure both source (pooling layer) and target (weight layer) have been previously processed
+    try:
+        source_layer_idx = node_2_layer_map[edge[0]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[0])
+    try:
+        destination_layer_idx = node_2_layer_map[edge[1]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[1])
+
+    # Find current source node within destinations
+    layer_info = dynapcnn_layer_info[source_layer_idx]
+    matched = False
+    for destination in layer_info["destinations"]:
+        if destination["pooling_ids"][-1] == edge[0]:
+            if destination["destination_layer"] is not None:
+                # Destination is already linked to a postsynaptic layer. This happens when
+                # pooling nodes have outgoing edges to different weight layer.
+                # Copy the destination
+                # TODO: Add unit test for this case
+                destination = {k: v for k, v in destination.items()}
```

Comment:
This seems strange to me. `destination` is already the variable being used in the for loop. In this line the `destination` of the for loop is being overwritten with its own content? Since .append is being used it seems to me that `layer_info['destinations']` is having its already existing content duplicated. 

Could we go over this on the next meeting @bauerfe ?

Reply:
Sure. To give a preliminary explanation:
It is correct that `destination` is being copied and appened again to the list of destinations. This is done if there are multiple edges to another dynapcnn layer destination  from a single pooling node. The way it is currently implemented, it will first only create one entry in `"destinations"` for this pooling node. However, it will need one for each destination layer that node is connected to. Hence the copying. I couldn't find a better way of setting this up. If you have a better idea that would be great. I though about including the edges to the subsequent weight layers in the call to `trace_paths`, but found that it wouldn't make it clearer.